src/ephemeris/shed_tools.py

```diff
@@ -232,75 +235,103 @@ def test_tools(self,
-
-    def _test_tool(self, tool, test_user, test_user_api_key):
-
+        with ThreadPoolExecutor(max_workers=parallel_tests) as executor:
+            try:
+                for tool in installed_tools:
+                    self._test_tool(executor=executor,
+                                    tool=tool,
+                                    galaxy_interactor=galaxy_interactor,
+                                    test_history=test_history,
+                                    log=log,
+                                    tool_test_results=all_test_results,
+                                    tests_passed=tests_passed,
+                                    test_exceptions=test_exceptions,
+                                    )
+            finally:
+                # Always write report, even if test was cancelled.
+                try:
+                    executor.shutdown(wait=True)
+                except KeyboardInterrupt:
+                    executor._threads.clear()
+                    thread._threads_queues.clear()
+                report_obj = {
+                    'version': '0.1',
+                    'tests': sorted(all_test_results, key=lambda el: el['id']),
+                }
+                with open(test_json, "w") as f:
+                    json.dump(report_obj, f)
+                log.info("Report written to '%s'", os.path.abspath(test_json))
+                if log:
+                    log.info("Passed tool tests ({0}): {1}".format(
+                        len(tests_passed),
+                        [t for t in tests_passed])
+                    )
+                    log.info("Failed tool tests ({0}): {1}".format(
+                        len(test_exceptions),
+                        [t[0] for t in test_exceptions])
+                    )
+                log.info("Total tool test time: {0}".format(dt.datetime.now() - tool_test_start))
+
+    def _get_interactor(self, test_user, test_user_api_key):
         if test_user_api_key is None:
             whoami = self.gi.make_get_request(self.gi.url + "/whoami").json()
             if whoami is not None:
                 test_user_api_key = self.gi.key
         galaxy_interactor_kwds = {
             "galaxy_url": re.sub('/api', '', self.gi.url),
             "master_api_key": self.gi.key,
-            "api_key": None,  # TODO
+            "api_key": test_user_api_key,  # TODO
             "keep_outputs_dir": '',
         }
         if test_user_api_key is None:
             galaxy_interactor_kwds["test_user"] = test_user
         galaxy_interactor = GalaxyInteractorApi(**galaxy_interactor_kwds)
+        return galaxy_interactor
+
+    @staticmethod
+    def _test_tool(executor,
+                   tool,
+                   galaxy_interactor,
+                   test_history=None,
+                   log=None,
+                   tool_test_results=None,
+                   tests_passed=None,
```

Comment:
I don't see `tests_passed` and `test_exceptions` defined. Yet the code appends to them whether they are lists. Shouldn't these be empty lists then by default?

Reply:
This is a private method that is the target function for the ThreadPoolExecutor, to be called from `test_tools`. I can make these non-default, which makes some sense. `tests_passed ` and `test_exceptions `  are defined in `test_tools`.

Reply:
:+1: Then better make them non-default indeed. The error will be less weird if anything goes wrong.